### PR TITLE
feat(eventProcessor): Add EventProcessor and BatchEventProcessor

### PR DESCRIPTION
**Pull request by mnoman09**
Summary
-------

- Introduces an EventProcessor interface.
- Introduces a BatchEventProcessor

Buffering events within a queue before dispatching is an optimization that should prevent SDK implementations from exhausting resources while increasing throughput. This implementation relies on a BlockingCollection to buffer events received from one-to-many producers. A single consumer thread continuously polls from this queue to build a batch before emitting the batched LogEvent.

Test plan
---------

- Added unit tests.

### CODE REVIEW

File: optimizely/closeable.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+
+ABC = abc.ABCMeta('ABC', (object,), {'__slots__': ()})
+
+
+class Closeable(object):
```
Comment (aliabbasrizvi) [inline optimizely/closeable.py]: Why do we need this class? I don't want us to impose object oriented programming.

File: optimizely/config_manager.py [hunk 2]
```diff
@@ -30,282 +30,282 @@
 
 
 class BaseConfigManager(ABC):
```
Comment (aliabbasrizvi) [inline optimizely/config_manager.py]: The 4 space indent was intentional. We want to move to that.

Please restore.

File: tox.ini [hunk 3]
```diff
@@ -4,6 +4,7 @@
 # E121 - continuation line indentation is not a multiple of four
 # E127 - continuation line over-indented for visual indent
 # E722 - do not use bare 'except'
-ignore = E111,E114,E121,E127, E722
+# W504 - line break after binary operator
```
Comment (aliabbasrizvi) [inline tox.ini]: Remove this and fix the indent problem. We are trying to clean up the remaining rules and become flake8 compliant.

File: optimizely/config_manager.py [hunk 4]
```diff
@@ -30,282 +30,282 @@
...
+     Args:
+       datafile: JSON string representing the Optimizely project.
+     """
+
+    if self.validate_schema:
+        if not validator.is_datafile_valid(datafile):
+            self.logger.error(enums.Errors.INVALID_INPUT.format('datafile'))
             return
 
-        self._config = config
```
Comment (aliabbasrizvi) [inline optimizely/config_manager.py]: How are config manager changes related to event processor changes? Please remove these changes from this PR.

File: optimizely/event/event_processor.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+class EventProcessor(ABC):
+  """ Class encapsulating event_processor functionality. Override with your own processor
+  providing process method. """
+
+  @abc.abstractmethod
+  def process(user_event):
+    pass
+
+
+class BatchEventProcessor(EventProcessor, Closeable):
```
Comment (aliabbasrizvi) [inline optimizely/event/event_processor.py]: So here is what I would recommend. 

You should define a timer class that does stuff based on a defined interval. I am hoping that the timer class will be useful in both config manager and here in the batch event processor.

Comment (oakbani) [reply optimizely/event/event_processor.py]: do you mean something similar to this https://github.com/optimizely/ruby-sdk/blob/master/lib/optimizely/config_manager/async_scheduler.rb

File: optimizely/event/event_processor.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+  def _validate_intantiation_props(self, prop, prop_name):
+    if (prop_name == 'batch_size' and not isinstance(prop, int)) or prop is None or prop < 1 or \
+      not validator.is_finite_number(prop):
+      self.logger.info('Using default value for {}.'.format(prop_name))
+      return False
+
+    return True
+
+  def _get_time(self, _time=None):
```
Comment (aliabbasrizvi) [inline optimizely/event/event_processor.py]: What is the purpose of this time? What is the unit? Can we add docstring?

File: optimizely/event/event_processor.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    if _time is None:
+      return int(round(time.time()))
+
+    return int(round(_time))
+
+  def start(self):
```
Comment (aliabbasrizvi) [inline optimizely/event/event_processor.py]: What is the purpose of this method? Can we add docstring?

File: optimizely/event/event_processor.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+
+    self.flushing_interval_deadline = self._get_time() + self._get_time(self.flush_interval.total_seconds())
+    self.executor = threading.Thread(target=self._run)
+    self.executor.setDaemon(True)
+    self.executor.start()
+
+    self._is_started = True
+
+  def _run(self):
+    """ Scheduler method that periodically flushes events queue. """
```
Comment (aliabbasrizvi) [inline optimizely/event/event_processor.py]: This is not entirely true right?

File: optimizely/event/event_processor.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    try:
+      while True:
+        if self._get_time() > self.flushing_interval_deadline:
+          self._flush_queue()
+
+        try:
+          item = self.event_queue.get(True, 0.05)
+
+        except queue.Empty:
+          self.logger.debug('Empty queue, sleeping for 50ms.')
```
Comment (aliabbasrizvi) [inline optimizely/event/event_processor.py]: nit. This can be a very loud message happening 20 times per second. I'd recommend not having this message.

**Pull request merged**